File: project-euler/euler-java/src/main/java/com/github/adrianomoreira/projecteuler100/Problem012.java
package com.github.adrianomoreira.projecteuler100;

public class Problem012 {


    static int getFactorsCount(long n) {
        int count = 2;
        var end = Math.sqrt(n);
        for (int i = 2; i < end; i++) {
            if (n % i == 0) {
                count += 2;
            }
        }
        if ((long) end * (long) end == n) {
            count++;
        }
        return count;
    }

    public static int divisibleTriangleNumber(long factorsCount) {
        int line = 1;
        int sum = 1;
        int currentFactorsCount;
        do {
            line++;
            sum += line;
            currentFactorsCount = getFactorsCount(sum);
        } while (currentFactorsCount - 1 < factorsCount);
        return sum;
    }
}

File: project-euler/euler-rust/src/problem012.rs
fn factors_count(n: u64) -> u64 {
    let mut count = 2;

    let end = (n as f64).sqrt() as u64;

    for i in 2..end {
        if n % i == 0 {
            count = count + 2;
        }
    }

    if end * end == n {
        count = count + 1;
    }

    count
}

fn divisible_triangle_number(n: u64) -> u64 {
    let mut i = 1;
    let mut sum = i;
    while factors_count(sum) < n {
        i = i + 1;
        sum = sum + i;
    }

    sum
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test() {
        assert_eq!(28, divisible_triangle_number(5));
        assert_eq!(630, divisible_triangle_number(23));
        assert_eq!(1385280, divisible_triangle_number(167));
        assert_eq!(17907120, divisible_triangle_number(374));
        assert_eq!(76576500, divisible_triangle_number(500));
    }
}